contracts/0.4.24/oracle/LidoOracle.sol

```diff
@@ -73,153 +76,211 @@ contract LidoOracle is ILidoOracle, AragonApp {
-
+        require(0 != _quorum, "QUORUM_WONT_BE_MADE");
         QUORUM_POSITION.setStorageUint256(_quorum);
         emit QuorumChanged(_quorum);
+        _tryPush(REPORTABLE_EPOCH_ID_POSITION.getStorageUint256());
+    }
 
-        uint256 minReportableEpochId = (
-            MIN_REPORTABLE_EPOCH_ID_POSITION.getStorageUint256()
-        );
-        uint256 lastReportedEpochId = (
-            LAST_REPORTED_EPOCH_ID_POSITION.getStorageUint256()
-        );
-
-        assert(lastReportedEpochId <= getCurrentEpochId());
-
-        if (lastReportedEpochId > minReportableEpochId) {
-            minReportableEpochId = lastReportedEpochId;
-            _tryPush(lastReportedEpochId);
-        }
+    /**
+     * @notice Returns the callback contract address to be called upon quorum
+     */
+    function getQuorumCallback() public view returns(address) {
+        return address(QUORUM_CALLBACK_POSITION.getStorageUint256());
+    }
 
-        _assertInvariants();
+    /**
+     * @notice Set the callback contract address to be called upon quorum
+     */
+    function setQuorumCallback(address _addr) external auth(SET_QUORUM_CALLBACK) {
+        QUORUM_CALLBACK_POSITION.setStorageUint256(uint256(_addr));
+        emit QuorumCallbackSet(_addr);
     }
 
     /**
      * @notice An oracle committee member reports data from the ETH 2.0 side
-     * @param _epochId BeaconChain epoch id
+     * @param _epochId Beacon Chain epoch id
      * @param _beaconBalance Balance in wei on the ETH 2.0 side
      * @param _beaconValidators Number of validators visible on this epoch
      */
     function reportBeacon(uint256 _epochId, uint128 _beaconBalance, uint128 _beaconValidators) external {
-        (uint256 startEpoch, uint256 endEpoch) = getCurrentReportableEpochs();
-        require(_epochId >= startEpoch, "EPOCH_IS_TOO_OLD");
+        BeaconSpec memory beaconSpec = _getBeaconSpec();
+        (uint256 currentEpoch, uint256 endEpoch) = getCurrentReportableEpochs();
+        require(_epochId >= currentEpoch, "EPOCH_IS_TOO_OLD");
         require(_epochId <= endEpoch, "EPOCH_HAS_NOT_YET_BEGUN");
+        require(_epochId.mod(beaconSpec.epochsPerFrame) == 0, "MUST_BE_FIRST_EPOCH_FRAME");
 
-        address member = msg.sender;
-        uint256 index = _getMemberId(member);
-        require(index != MEMBER_NOT_FOUND, "MEMBER_NOT_FOUND");
+        // If reported epoch has advanced, clear the last unsuccessful reporting
+        if (_epochId > currentEpoch) _clearReportingAndAdvanceTo(_epochId);
 
-        // check & set contribution flag
-        uint256 bitMask = gatheredEpochData[_epochId].reportsBitMask;
+        // Make sure the oracle is from members list and has not yet voted
+        uint256 index = _getMemberId(msg.sender);
+        require(index != MEMBER_NOT_FOUND, "MEMBER_NOT_FOUND");
+        uint256 bitMask = REPORTS_BITMASK_POSITION.getStorageUint256();
         require(!bitMask.getBit(index), "ALREADY_SUBMITTED");
```

Comment:
It's better to merge `getBit` and `setBit` to one `testAndSet` method

Reply:
Fix in https://github.com/lidofinance/lido-dao/pull/311.